### PR TITLE
[BEAM-6079] Add ability for CassandraIO to delete data

### PR DESCRIPTION
**Pull request by fabienrousseau**
Currently, it's possible to read & write data using CassandraIO.

It would be nice to be able to delete using a CassandraIO.delete().

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

File: sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java [hunk 1]
```diff
@@ -499,4 +506,191 @@ public void teardown() throws Exception {
       writer = null;
     }
   }
+
+  /**
+   * A {@link PTransform} to delete into Apache Cassandra. See {@link CassandraIO} for details on
+   * usage and configuration.
+   */
+  @AutoValue
+  public abstract static class Delete<T> extends PTransform<PCollection<T>, PDone> {
```
Comment (iemejia) [inline sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java]: Do you think we can reuse (maybe via inheriting from Write or making a new class to config both) the configuration from `Write`). I am a bit uneasy with the duplicate code.

File: sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java [hunk 2]
```diff
@@ -471,4 +471,83 @@ private void waitForFuturesToFinish() throws ExecutionException, InterruptedExce
...
+    /**
+     * Delete the entity to the Cassandra instance, using {@link Mapper} obtained with the {@link
+     * MappingManager}. This method uses {@link Mapper#deleteAsync(Object)} method, which is
+     * asynchronous. Beam will wait for all futures to complete, to guarantee all writes have
+     * succeeded.
+     */
+    @Override
+    public void delete(T entity) throws ExecutionException, InterruptedException {
+      Mapper<T> mapper = (Mapper<T>) mappingManager.mapper(entity.getClass());
+      this.deleteFutures.add(mapper.deleteAsync(entity));
```
Comment (iemejia) [inline sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java]: Similar to the previous comment. Do you think we can make current Writer somehow extensible for this case? I have the impression this line is the only difference between the write and the delete, or I am missing something else?

File: sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java (hunk 1 above)
Comment (fabienrousseau) [reply sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java]: Sure we can probably reuse. This was a first version to have something working quickly and start a discussion.
Thanks for the feedback. I'll improve the patch soon.

**Comment by fabienrousseau** [comment]
Hi,

The new commit avoid code duplication by introducing a new Mutate class (which can be parameterized to either write or delete data).

**Comment by iemejia** [comment]
Merged manually, thanks again !

**Comment by fabienrousseau** [comment]
Awesome !
Thanks

**Pull request closed without merging**